### PR TITLE
[RTM] Feature/remove html cache

### PR DESCRIPTION
**Pull request by Toflar**
This is a spin-off PR of #562. It removes our own html cache and adds support for caching via proxies.

To make this work, you need to adjust your `config.yml`:

```diff
# Framework configuration
framework:
    ...
    translator: { fallbacks: ["%locale%"] }
+   esi: { enabled: true }
+   fragments: { enabled: true, path: /_fragment }
```

### CODE REVIEW

File: src/EventListener/AddToSearchIndexListener.php [hunk 1]
```diff
@@ -48,6 +48,12 @@ public function onKernelTerminate(PostResponseEvent $event)
             return;
         }
 
+        // Hard coded support for _fragment URIs that should not get indexed.
+        // TODO: remove this listener completely and work towards a real crawler.
+        if (preg_match('/_fragment/', $event->getRequest()->getPathInfo())) {
```
Comment (aschempp) [inline src/EventListener/AddToSearchIndexListener.php]: Maybe the hardcoded `_fragment` could be injected from the container?

Comment (Toflar) [inline src/EventListener/AddToSearchIndexListener.php]: I could, yes. Just didn't do because it's going to be dropped anyway.

Comment (Toflar) [inline src/EventListener/AddToSearchIndexListener.php]: Added in 7f13a07b9a45ac993865f6f933ec945bc08f9d41.

**Comment by discordier** [comment]
@Toflar Would you fix the unit tests or do you want a PR against your PR? :)

**Comment by Toflar** [comment]
I'll fix them once we decided how to proceed on the caching issues. /cc @leofeyer

**Comment by discordier** [comment]
I was just asking because it was labeled as RTM. :)

**Comment by leofeyer** [comment]
The problem is that pages with a mobile layout cannot be cached anymore if we merge this.

It will not affect me, because I never use mobile layouts, but I cannot estimate how many people are using the feature. 

Also, I understand that having two different layouts for the same page can never be handled by any kind of reverse proxy, so this seems like a feature we should deprecate and remove in Contao anyway.

**Comment by Toflar** [comment]
> Also, I understand that having two different layouts for the same page can never be handled by any kind of reverse proxy, so this seems like a feature we should deprecate and remove in Contao anyway.

It can be handled if we provide the means to find out what header to vary on based on the cookie. I'll work on this because we need it for authenticated members anyway. Nevertheless, this feature should be removed anyway in Contao 5, I agree :)

**Comment by Toflar** [comment]
Okay. This PR should be RTM then. The only difference to the previous implementation is that currently the mobile page layout cannot be cached by the proxy. I'll work on this though. On the other hand we can now use real proxies such as Varnish and also benefit by ESI implementations in our own application code which is a huge advantage in my opinion.

**Comment by aschempp** [comment]
If you cannot cache a mobile layout, how do you cache a desktop layout? Or do you mean caching is completly disabled if there is a mobile layout assigned to a page?

**Comment by Toflar** [comment]
> Or do you mean caching is completly disabled if there is a mobile layout assigned to a page?

Exactly.

**Comment by leofeyer** [comment]
Since we are not getting ESI in Contao 4.3, do we still want to merge this PR?

**Comment by Toflar** [comment]
Of course! That's why it's a spin-off PR...

**Comment by aschempp** [comment]
Absolutely, storing the output in the proxy cache would still be way better than our current listener.

**Comment by aschempp** [comment]
I don't think this is RTM though. There are still a few things I'd do differently, from the controller name to the question of data upgrade. Not sure about deadlines though?

**Comment by discordier** [comment]
@aschempp How about writing down what controller name you would like better?

File: src/Controller/EsiController.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Handles the Contao ESI requests. This might be subject to change
+ * in the very near future which is why this class is declared final.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+final class EsiController extends Controller
```
Comment (aschempp) [inline src/Controller/EsiController.php]: This should rather be `InsertTagsController`

File: src/Controller/EsiController.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $this->framework = $framework;
+    }
+
+    /**
+     * Renders non cacheable insert tags (such as {{request_token}}).
+     *
+     * @param string $insertTag
+     */
+    public function renderNonCacheableInsertTag($insertTag)
```
Comment (aschempp) [inline src/Controller/EsiController.php]: which would make this `indexAction`?

File: src/EventListener/AddToSearchIndexListener.php [hunk 4]
```diff
@@ -48,6 +55,10 @@ public function onKernelTerminate(PostResponseEvent $event)
             return;
         }
 
+        if (preg_match('/' . preg_quote($this->fragmentPath, '/') .'/', $event->getRequest()->getPathInfo())) {
```
Comment (aschempp) [inline src/EventListener/AddToSearchIndexListener.php]: Maybe that should be checked in `indexPageIfApplicable`, that's what the function's name says it does?

File: src/Resources/contao/classes/Frontend.php [hunk 5]
```diff
@@ -654,224 +654,16 @@ public static function indexPageIfApplicable(Response $objResponse)
...
-			$response->headers->set('Last-Modified', gmdate('D, d M Y H:i:s', time()) . ' GMT');
-			$response->headers->set('Expires', gmdate('D, d M Y H:i:s', $expire) . ' GMT');
-		}
-		else
-		{
-			$response->headers->set('Cache-Control', 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0');
-			$response->headers->set('Last-Modified', gmdate('D, d M Y H:i:s') . ' GMT');
-			$response->headers->set('Expires', 'Fri, 06 Jun 1975 15:10:00 GMT');
-		}
+		@trigger_error('Using Frontend::getResponseFromCache() has been deprecated and will no longer work in Contao 5.0.', E_USER_DEPRECATED);
```
Comment (aschempp) [inline src/Resources/contao/classes/Frontend.php]: I think the `trigger_error` message should also include the information that the method does not work anymore (like the docblock header).

File: src/Resources/contao/classes/FrontendTemplate.php [hunk 6]
```diff
@@ -241,98 +240,13 @@ public function hasAuthenticatedBackendUser()
 
 	/**
 	 * Add the template output to the cache and add the cache headers
+	 *
+	 * @deprecated Deprecated since Contao 4.3, to be removed in Contao 5.0.
```
Comment (aschempp) [inline src/Resources/contao/classes/FrontendTemplate.php]: Again, it's not only deprecated, it no longer does anything at all.

File: src/Resources/contao/classes/FrontendTemplate.php [hunk 7]
```diff
@@ -414,4 +328,39 @@ public function getCustomSections($strKey=null)
...
+		global $objPage;
+
+		if (false === $objPage->cache && false === $objPage->clientCache)
+		{
+			return;
+		}
+
+		// If FE_USER_LOGGED_IN or BE_USER_LOGGED_IN every request is private
+		// Moreover, mobile layout and protected pages are not cached either
+		// TODO: Add support for proxies so they can vary on member context and page layout
```
Comment (aschempp) [inline src/Resources/contao/classes/FrontendTemplate.php]: If I remember our discussion correctly, that TODO can be removed?

File: src/Resources/contao/classes/FrontendTemplate.php [hunk 8]
```diff
@@ -414,4 +328,39 @@ public function getCustomSections($strKey=null)
...
+		if (true === FE_USER_LOGGED_IN || true === BE_USER_LOGGED_IN || true === $objPage->isMobile || $objPage->protected)
+		{
+			$response->setPrivate();
```
Comment (aschempp) [inline src/Resources/contao/classes/FrontendTemplate.php]: The method should set the response as private initially (e.g. around line 341) and then just `return` in the necessary cases. `setSharedMaxAge` will set it to *public* and remove the *private* header.

File: src/Controller/EsiController.php (hunk 3 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: No, it still only renders the non cacheable ones.

File: src/Controller/EsiController.php (hunk 2 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: Jup.

File: src/EventListener/AddToSearchIndexListener.php (hunk 4 above)
Comment (Toflar) [inline src/EventListener/AddToSearchIndexListener.php]: No. I think it's better suited here.

File: src/Resources/contao/classes/Frontend.php (hunk 5 above)
Comment (Toflar) [inline src/Resources/contao/classes/Frontend.php]: I can add that, yes.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 6 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Agreed.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 7 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: No, that TODO still remains.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 8 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Disagree. It's more explicit like that. The other way would require one to know that `setSharedMaxAge` does reset the private header.

File: src/Controller/EsiController.php (hunk 3 above)
Comment (aschempp) [inline src/Controller/EsiController.php]: Well, not true, it renders **all** insert tags **including** cacheable ones. Also, shouldn't an action have `Action` suffix?

File: src/Resources/contao/models/PageModel.php [hunk 9]
```diff
@@ -927,12 +930,6 @@ public function loadDetails()
 
 		$this->trail = array_reverse($trail);
 
-		// Do not cache protected pages
-		if ($this->protected)
```
Comment (leofeyer) [inline src/Resources/contao/models/PageModel.php]: Why did you remove this?

File: src/Controller/EsiController.php (hunk 2 above)
Comment (leofeyer) [inline src/Controller/EsiController.php]: Declaring a class `final` is not the appropriate thing to do if it is subject to change. We should rather declare it `@internal` instead, shouldn't we?

File: src/Resources/contao/classes/FrontendTemplate.php [hunk 10]
```diff
@@ -124,15 +127,11 @@ protected function compile($blnCheckRequest=false)
...
-		$this->addToCache();
-
 		// Unset only after the output has been cached (see #7824)
 		unset($_SESSION['LOGIN_ERROR']);
 
-		// Replace insert tags and then re-replace the request_token tag in case a form element has been loaded via insert tag
-		$this->strBuffer = $this->replaceInsertTags($this->strBuffer, false);
-		$this->strBuffer = str_replace(array('{{request_token}}', '[{]', '[}]'), array(REQUEST_TOKEN, '{{', '}}'), $this->strBuffer);
+		// Replace insert tags
+		$this->strBuffer = $this->replaceInsertTags($this->strBuffer);
```
Comment (leofeyer) [inline src/Resources/contao/classes/FrontendTemplate.php]: Why did you remove the second argument here?

File: src/Controller/EsiController.php (hunk 3 above)
Comment (leofeyer) [inline src/Controller/EsiController.php]: They should.

**Comment by leofeyer** [comment]
I will merge this as soon as we have discussed @aschempp's comments.

File: src/Controller/EsiController.php (hunk 2 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: Well it shouldn't be subject to change anymore anyway as it's only handling insert tags. I'll remove it.

File: src/Controller/EsiController.php (hunk 3 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: I'll call it `renderAction`.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 10 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Because I want it to be `true` so they are replaced by the `<esi>` tags and not ignored.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 8 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Does it? Hm, okay, will look into it.

File: src/Resources/contao/models/PageModel.php (hunk 9 above)
Comment (Toflar) [inline src/Resources/contao/models/PageModel.php]: Because it's wrong and not needed. Caching should not be disabled just because the page is protected. In this PR, whenever a user is logged in, the response is set to private, so it never caches.

File: src/Controller/EsiController.php (hunk 2 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: Implemented in 071d434e163b9b8924cf90048b33c2838d02171a.

File: src/Controller/EsiController.php (hunk 3 above)
Comment (Toflar) [inline src/Controller/EsiController.php]: Implemented in 071d434e163b9b8924cf90048b33c2838d02171a.

File: src/Resources/contao/classes/Frontend.php (hunk 5 above)
Comment (Toflar) [inline src/Resources/contao/classes/Frontend.php]: Added in d5f3bf53d18b787c676c18fa4daa11811b8ac9e6.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 6 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Added in d5f3bf53d18b787c676c18fa4daa11811b8ac9e6.

File: src/Resources/contao/classes/FrontendTemplate.php (hunk 8 above)
Comment (Toflar) [inline src/Resources/contao/classes/FrontendTemplate.php]: Implemented explicit private header in 69cc0223ebd61fcd2a982cea6e82cc93e10af3c6. A response is private by default anyway.

**Comment by Toflar** [comment]
All review comments addressed.

**Comment by discordier** [comment]
All good from my side also. Big thanks to @Toflar for making this happen.
